### DISCUSSION TITLE
[nvim-lspconfig] Should I expcect `plugins.extras.clangd` to work out of the box on MacOS?

### DISCUSSION THREAD

**Discussion by alexsanduk** [LazyVim/LazyVim#2124] [Q&A] [answered]
I am running `Sonoma 14.1.1`. I have enabled `clangd` plugin in `lazy.lua`:
```
    { import = "lazyvim.plugins.extras.lang.json" },
    { import = "lazyvim.plugins.extras.lang.clangd" },
    { import = "lazyvim.plugins.extras.lang.python" },
    { import = "lazyvim.plugins.extras.ui.mini-animate" },
    -- import/override with your plugins
    { import = "plugins" },
```
I don't get autocompletion for std libs:
![Screenshot 2023-12-04 at 16 51 45](https://github.com/LazyVim/LazyVim/assets/6910120/65b9e93f-8601-40af-8861-f0d8ae3a85fa)
`LspInfo`
![Screenshot 2023-12-04 at 16 55 56](https://github.com/LazyVim/LazyVim/assets/6910120/c86817aa-776e-412d-8811-aaa16db9519c)

I was expecting it to work out of the box. Is it related to mason installing non-system `clangd`?

**Comment by dpetka2001** [comment]
Which version of clang do you have installed on your system?

**Reply by alexsanduk** [reply]
Mason version:
```
clangd version 17.0.3 (https://github.com/llvm/llvm-project 888437e1b60011b8a375dd30928ec925b448da57)
Features: mac+grpc+xpc
Platform: arm64-apple-darwin23.1.0; target=x86_64-apple-darwin23.1.0
```
Apple version:
```
Apple clangd version 15.0.0 (clang-1500.0.40.1)
Features: mac+xpc
Platform: x86_64-apple-darwin23.1.0; target=arm64-apple-darwin23.1.0
```

**Reply by dpetka2001** [reply]
This is the `clangd` version. I'm asking about `clang` the compiler itself. Do you have it installed?

**Reply by dpetka2001** [reply]
Never mind, I didn't read properly. What does `clang-15 -v` show?

**Reply by alexsanduk** [reply]
Sorry, right `clang -v`:
```
Apple clang version 15.0.0 (clang-1500.0.40.1)
Target: arm64-apple-darwin23.1.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Reply by dpetka2001** [reply]
For me it shows the following
```
Ubuntu clang version 15.0.7
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/10
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/11
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/12
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/12
```
Do you have `gcc` installed?

**Reply by dpetka2001** [reply]
Do you have an executable `clang-15`? Does `clang-15 -v` show a different output from what you already showed?

**Reply by dpetka2001** [reply]
Essentially you need to install the development system headers on your OS. From clangd [docs](https://clangd.llvm.org/guides/system-headers#headers-not-present-in-the-system-at-all) it seems that on Mac you need to install 
> libc++ or libstdc++ for mac, either through brew or XCode

Try that and see if it works for you.

**Reply by alexsanduk** [reply]
I only have the `clang-15` that was installed along with command line tools. I haven't installed `gcc`. 
Now I am wondering if completely avoiding system `clang` may resolve the issue. 
I am just wondering what is out-of-the-box solution for macOS

**Reply by dpetka2001** [reply]
Pls see my previous comment and check if it works for you.

**Reply by alexsanduk** [reply, marked as answer]
The issue isn't related to missing libraries, but rather mason `clangd` not knowing where to search. Therefore better to skip mason installing `clangd`:
```lua
return {
  "neovim/nvim-lspconfig",
  opts = {
    setup = {
      clangd = function(_, opts)
        opts.capabilities.offsetEncoding = { "utf-16" }
      end,
    },
    servers = {
      clangd = { mason = false },
    },
  },
}

```
This way it works with system `clangd`

**Reply by dpetka2001** [reply]
Glad you got it working. I also did have this problem at the beginning but as soon as I installed the system headers on my OS (libstdc++), mason `clangd` was able to find them, since they were installed in my PATH. I don't have any problem with `clangd` installed by mason. It might also have something to do with how things are configured on Mac maybe, I definitely can't tell for sure. But on Linux distros, I've seen many people solving this problem by installing the necessary `libstdc++` development headers.
